feat(bigquery): Custom message when Service Account doesnt have the correct Roles and Permissions

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
`BigQuery` is now returning the project name twice in the error message when connecting, which breaks the Regex in the `custom_errors` dictionary, so we need to adjust it in order to stop displaying the general connection failed message. Also, we are updating the message displayed to users and the tests.

Additionally, in order to actually being able to catch the right error, we are raising the exceptions with their original message instead of the default one, otherwise the exception handler will ignore it and display the default message.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
![error](https://user-images.githubusercontent.com/38889534/196283269-38a6f706-fadc-4282-837d-be0d2371b5ce.gif)

After:
![test](https://user-images.githubusercontent.com/38889534/196283298-900b4552-3690-4dba-a9f1-fc245e14ec1f.gif)


### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [X] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
BigQuery:

    - Start sending the message from test connection exceptions instead of a generic exception with no message in it
    - Update the regex CONNECTION_DATABASE_PERMISSIONS_REGEX because google is sending the project name twice now
    - Update tests

Antonio-RiveroMartnez committed:
BigQuery:

- Add the original message ONLY for CONNECTION_DATABASE_PERMISSIONS_ERROR error types so we can keep using the default message for other cases used in other places like tests

hughhhh commented on superset/databases/commands/create.py:
```diff
@@ -46,6 +48,34 @@ def run(self) -> Model:
+            # so we try to return the message from the errors instead of using the
+            # default message
+            error_list: List[SupersetError] = ex.errors
+            for error in error_list:
```
instead of doing a loop can we just do set check?

`if SupersetErrorType.CONNECTION_DATABASE_PERMISSIONS_ERROR in error_list`

Antonio-RiveroMartnez replied:
Hmm `error_list` contains a list of `SupersetError`, while `SupersetErrorType.CONNECTION_DATABASE_PERMISSIONS_ERROR` is just a enum, that comparison won't work. Reducing error_list would also require iterating over the list, so, let me see if there's a more "Pythonic" way of checking without looping.

hughhhh commented on superset/databases/commands/create.py:
```diff
@@ -46,6 +48,34 @@ def run(self) -> Model:
+                    error.error_type
+                    == SupersetErrorType.CONNECTION_DATABASE_PERMISSIONS_ERROR
+                ):
+                    raise DatabaseConnectionFailedError(
```
can we just raise the original `DatabaseTestConnectionFailedError`

Antonio-RiveroMartnez replied:
Updated. I am catching the new exception type in the API now along with sending back to the user a dict with the error messages instead of just raw strings.

hughhhh replied:
can you show me the diff between the payload coming back from the API when you send `DatabaseConnectionFailedError` vs the original exception?

Antonio-RiveroMartnez replied:
Before I was raising `DatabaseConnectionFailedError` which only has one error in it, so the API response would look like:
`{message: '<message_here>'}` while now I'm just throwing the original exception (`DatabaseTestConnectionFailed`) which has a list of errors, so the API would respond with a dict like: `{[error_type]: [error_message], [error_type_2]: [error_message_2], ...}` which is already a known pattern to the `DatabaseModal` for error displaying.

So basically after the change based on the input you gave me, I'm doing the latter.

For example:

Before when throwing a `DatabaseConnectionFailedError` the API would respond:

Antonio-RiveroMartnez committed:
BigQuery:

- Raise the original DatabaseTestConnectionFailedError exception instead of a new one.
- Adding normalized_messages method in SupersetErrorsException since the to_dict is not useful for getting the right message to users

jinghua-qa reviewed (approved):
Able to see the custom msg for error msg. LGTM

Antonio-RiveroMartnez committed:
BigQuery:

- Raise the original DatabaseConnectionFailedError from the test command with the right message when necessary
- Remove the unnecessary code changes to handle multiple errors or DatabaseTestConnectionFailedError exceptions

Antonio-RiveroMartnez committed:
BigQuery:

        - Send a SupersetErrorsException instead of Test one since in the future we will be raising just superset exceptions
        - Update testsq

eschutho commented on superset/databases/commands/test_connection.py:
```diff
@@ -145,7 +152,7 @@ def ping(engine: Engine) -> bool:
             # check for custom errors (wrong username, wrong password, etc)
             errors = database.db_engine_spec.extract_errors(ex, context)
-            raise DatabaseTestConnectionFailedError(errors) from ex
+            raise SupersetErrorsException(errors) from ex
```
@Antonio-RiveroMartnez I'm looking into some 500 errors for test connections, and I saw this change. What would happen if I were to change this back to raise the `DatabaseTestConnectionFailedError` which is a 422? Would it break anything from this PR? I checked through the implementation and it looks like it should still catch the SupersetErrorsExceptions because it is the base class, but I wanted to see if I might be missing anything here.

Antonio-RiveroMartnez replied:
Hey @eschutho, the problem of doing that is that we would lost the `original message` and instead the message from DatabaseTestConnectionFailedError would be used (`Connection failed, please check your connection settings`) and not the custom message with roles after parsing the original message.
